packages/ckeditor5-bookmark/src/bookmarkediting.ts

```diff
@@ -181,27 +180,94 @@ export default class BookmarkEditing extends Plugin {
+ * A helper function to match an `anchor` element which must contain `id` or `name` attribute but without `href` attribute,
+ * also when `expectEmpty` is set to `true` but the element is not empty matcher should not match any element.
  *
  * @param element The element to be checked.
- * @param dataPipeline When set to `true` matcher is executed in data pipeline, checks if `element` is empty;
+ * @param expectEmpty Default set to `true`, when set to `false` matcher expects that `anchor` is not empty;
  * in editing pipeline it's not empty because it contains the `UIElement`.
  */
-function upcastMatcher( element: ViewElement, dataPipeline: boolean = true ) {
+function upcastMatcher( element: ViewElement, expectEmpty: boolean = true ) {
 	const isAnchorElement = element.name === 'a';
 
 	if ( !isAnchorElement ) {
 		return null;
 	}
 
+	if ( expectEmpty && !element.isEmpty ) {
+		return null;
+	}
+
 	const hasIdAttribute = element.hasAttribute( 'id' );
+	const hasNameAttribute = element.hasAttribute( 'name' );
 	const hasHrefAttribute = element.hasAttribute( 'href' );
-	const isEmpty = element.isEmpty;
 
-	if ( !hasIdAttribute || hasHrefAttribute || dataPipeline && !isEmpty ) {
-		return null;
+	if ( hasIdAttribute && !hasHrefAttribute ) {
+		return { name: true, attributes: [ 'id' ] };
+	}
+
+	if ( hasNameAttribute && !hasHrefAttribute ) {
+		return { name: true, attributes: [ 'name' ] };
 	}
 
-	return { name: true, attributes: [ 'id' ] };
+	return null;
+}
+
+/**
+ * A view-to-model converter that handles converting pointed or wrapped anchors with `id` and/or `name` attributes.
+ *
+ * @returns Returns a conversion callback.
+ */
+function dataViewModelAnchorInsertion( editor: Editor ) {
+	return (
+		evt: EventInfo,
+		data: UpcastConversionData<ViewElement>,
+		conversionApi: UpcastConversionApi
+	) => {
+		const viewItem = data.viewItem;
+		const match = upcastMatcher( viewItem, false );
+
+		if ( !match || !conversionApi.consumable.test( viewItem, match ) ) {
+			return;
+		}
+
+		const enableNonEmptyBookmarkConversion = isEnabledNonEmptyBookmarkConversion( editor );
+
+		if ( !enableNonEmptyBookmarkConversion && !viewItem.isEmpty ) {
+			return;
+		}
+
+		const modelWriter = conversionApi.writer;
+		const anchorId = viewItem.getAttribute( 'id' );
+		const anchorName = viewItem.getAttribute( 'name' );
+		const bookmarkId = anchorId || anchorName;
+		const bookmark = modelWriter.createElement( 'bookmark', { bookmarkId } );
+
+		if ( !conversionApi.safeInsert( bookmark, data.modelCursor ) ) {
+			return;
+		}
+
+		conversionApi.consumable.consume( viewItem, match );
+
+		if ( anchorId === anchorName ) {
+			conversionApi.consumable.consume( viewItem, { attributes: [ 'name' ] } );
+		}
+
+		conversionApi.updateConversionResult( bookmark, data );
+
+		const { modelCursor, modelRange } = conversionApi.convertChildren( viewItem, data.modelCursor );
```

Comment:
Let's add a code comment here. This part uses the result of bookmark insertion to convert the anchor content after the bookmark element (not inside it).